frostsnap_embedded/src/partition.rs

```diff
@@ -0,0 +1,324 @@
+        assert_eq!(S::ERASE_SIZE % BUFFER_SIZE, 0);
+        BincodeFlashWriter {
+            flash: *self,
+            buf: [0xff; BUFFER_SIZE],
+            buf_index: 0,
+            word_pos: 0,
+        }
+    }
+}
+
+pub struct BincodeFlashReader<'a, S> {
+    flash: FlashPartition<'a, S>,
+    pos: u32,
+}
+
+impl<S> BincodeFlashReader<'_, S> {
+    pub fn seek_byte(&mut self, pos: u32) {
+        self.pos = pos;
+    }
+
+    pub fn byte_pos(&self) -> u32 {
+        self.pos
+    }
+}
+
+impl<S: NorFlash> bincode::de::read::Reader for BincodeFlashReader<'_, S> {
+    fn read(&mut self, bytes: &mut [u8]) -> Result<(), bincode::error::DecodeError> {
+        // this only works because we're using the "bytewise-read" feature of esp-storage
+        self.flash.read(self.pos, bytes).map_err(|e| {
+            bincode::error::DecodeError::OtherString(format!("Flash read error {:?}", e))
+        })?;
+        self.pos += bytes.len() as u32;
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BincodeFlashWriter<'a, S, const BUFFER_SIZE: usize> {
+    flash: FlashPartition<'a, S>,
+    buf: [u8; BUFFER_SIZE],
+    buf_index: usize,
+    word_pos: u32,
+}
+
+impl<S: NorFlash, const BUFFER_SIZE: usize> BincodeFlashWriter<'_, S, BUFFER_SIZE> {
+    pub fn seek_word(&mut self, word: u32) {
+        self.word_pos = word;
+    }
+
+    pub fn curr_word(&self) -> u32 {
+        self.word_pos
+    }
+
+    pub fn flush(mut self) -> Result<u32, NorFlashErrorKind> {
+        if self.buf_index != 0 {
+            let aligned_index = self.buf_index
+                + ((S::WRITE_SIZE - (self.buf_index % S::WRITE_SIZE)) % S::WRITE_SIZE);
+            self.buf[self.buf_index..aligned_index].fill(0xff);
+            // set to zero so we don't get the drop panic even if we fail
+            self.buf_index = 0;
+            self.flash
+                .nor_write(
+                    self.word_pos * S::WRITE_SIZE as u32,
+                    &self.buf[..aligned_index],
+                )
+                .map_err(|e| e.kind())?;
+            self.word_pos += aligned_index as u32 / S::WRITE_SIZE as u32;
+        }
+
+        Ok(self.word_pos)
+    }
+}
+
+impl<S: NorFlash, const BUFFER_SIZE: usize> bincode::enc::write::Writer
+    for BincodeFlashWriter<'_, S, BUFFER_SIZE>
+{
+    fn write(&mut self, bytes: &[u8]) -> Result<(), bincode::error::EncodeError> {
+        let mut i = 0;
+        loop {
+            if i == bytes.len() {
+                break;
+            }
+
+            self.buf[self.buf_index] = bytes[i];
+            self.buf_index += 1;
+
+            if self.buf_index == BUFFER_SIZE {
+                self.flash
+                    .nor_write(self.word_pos * S::WRITE_SIZE as u32, &self.buf[..])
+                    .map_err(|e| bincode::error::EncodeError::OtherString(format!("{:?}", e)))?;
+                self.buf_index = 0;
+                self.word_pos += (BUFFER_SIZE / S::WRITE_SIZE) as u32;
```

Comment:
BUFFER_SIZE is always a multiple of WRITE_SIZE?

Reply:
yes. 

![image](https://github.com/user-attachments/assets/8d5bcbc2-7fbd-4856-8a7a-9179c95b7673)